luigi/parameter.py

```diff
@@ -163,31 +163,31 @@ def _get_value_from_config(self, section, name):
         return self.parse(value)
 
     def _get_value(self, task_name=None, param_name=None):
-        if self.__global != _no_value:
-            return self.__global
+        for value, warn in self._value_iterator(task_name, param_name):
+            if value != _no_value:
+                if warn:
+                    warnings.warn(warn, DeprecationWarning, stacklevel=2)
+                return value
+        return _no_value
+
+    def _value_iterator(self, task_name, param_name):
+        """
+        Yield the parameter values, with optional deprecation warning as second tuple value.
+
+        The parameter value will be whatever non-_no_value that is yielded first.
+        """
+        yield (self.__global, None)
         if task_name and param_name:
-            v = self._get_value_from_config(task_name, param_name)
-            if v != _no_value:
-                return v
-            v = self._get_value_from_config(task_name, param_name.replace('_', '-'))
-            if v != _no_value:
-                warnings.warn(
-                    'The use of the configuration [%s] %s (with dashes) should be avoided. Please use underscores.' %
-                    (task_name, param_name), DeprecationWarning, stacklevel=2)
-                return v
+            yield (self._get_value_from_config(task_name, param_name), None)
+            warn = 'Configuration [{}] {} (with dashes) should be avoided. Please use underscores.'.format(
+                   task_name, param_name)
+            yield (self._get_value_from_config(task_name, param_name.replace('_', '-')), None)
         if self.__config:
-            v = self._get_value_from_config(self.__config['section'], self.__config['name'])
-            if v != _no_value and task_name and param_name:
-                warnings.warn(
-                    'The use of the configuration [%s] %s is deprecated. Please use [%s] %s' %
-                    (self.__config['section'], self.__config['name'], task_name, param_name),
-                    DeprecationWarning, stacklevel=2)
-            if v != _no_value:
-                return v
-        if self.__default != _no_value:
-            return self.__default
-
-        return _no_value
+            warn = (task_name and param_name and
+                    'The use of the configuration [{}] {} is deprecated. Please use [{}] {}'.format(
+                    self.__config['section'], self.__config['name'], task_name, param_name))
```

Comment:
The constructor of `Parameter()` or of `Task()`? Currently this code path will be executed when you instantiate a task (the constructor of `Task()`). I suppose it could be in the parameter's constructor too I guess. But I'm actually striving for making a `Parameter()`-object only data without side effects. I wouldn't like creating a `Parameter()` to have side effects.